Source/C++/Core/Ap4FileWriter.cpp

```diff
@@ -76,6 +78,47 @@ AP4_FileWriter::Write(AP4_File& file, AP4_ByteStream& stream, Interleaving /* in
     unsigned int t=0;
     AP4_Result result = AP4_SUCCESS;
     AP4_UI64   mdat_size = AP4_ATOM_HEADER_SIZE;
+    // compute mdat_size
+    for (AP4_List<AP4_Track>::Item* track_item = movie->GetTracks().FirstItem();
+         track_item; track_item = track_item->GetNext()) {
+        AP4_Track*    track = track_item->GetData();
+        AP4_Cardinal     sample_count = track->GetSampleCount();
+        AP4_SampleTable* sample_table = track->GetSampleTable();
+        AP4_Sample       sample;
+        for (AP4_Ordinal i=0; i<sample_count; i++) {
+            sample_table->GetSample(i, sample);
+            mdat_size += sample.GetSize();
+        }
+    }
+
+    // replace stco with co64
+    if (mdat_size >> 32) {
+        for (AP4_List<AP4_Track>::Item* track_item = movie->GetTracks().FirstItem();
+             track_item; track_item = track_item->GetNext()) {
+            AP4_Track*    track = track_item->GetData();
+            AP4_TrakAtom* trak  = const_cast<AP4_TrakAtom*>(track->GetTrakAtom());
+            AP4_StcoAtom* stco = AP4_DYNAMIC_CAST(AP4_StcoAtom, trak->FindChild("mdia/minf/stbl/stco"));
+            if (stco) {
+                AP4_Cardinal chunk_count = stco->GetChunkCount();
+                AP4_UI64* chunk_offsets = new AP4_UI64[chunk_count];
+                AP4_UI32* chunk_offsets_32 = stco->GetChunkOffsets();
+                for (unsigned int i=0; i<chunk_count; i++) {
+                    chunk_offsets[i] =  chunk_offsets_32[i];
+                }
+                AP4_Co64Atom* co64 = new AP4_Co64Atom(chunk_offsets, chunk_count);
+                stco->Detach();
+                AP4_ContainerAtom* stbl = AP4_DYNAMIC_CAST(AP4_ContainerAtom, trak->FindChild("mdia/minf/stbl"));
+                if (stbl == NULL) {
+                	return AP4_ERROR_INTERNAL;
```

Comment:
I saw the FIXME comment also and saw this pull request. You will be leaking co64 here when you return.